lib/spack/spack/build_environment.py

```diff
@@ -803,7 +804,94 @@ def modifications_from_dependencies(spec, context):
     return env
 
 
-def fork(pkg, function, dirty, fake):
+def _setup_pkg_and_run(serialized_pkg, function, kwargs, child_pipe,
+                       input_multiprocess_fd):
+
+    try:
+        # We are in the child process. Python sets sys.stdin to
+        # open(os.devnull) to prevent our process and its parent from
+        # simultaneously reading from the original stdin. But, we assume
+        # that the parent process is not going to read from it till we
+        # are done with the child, so we undo Python's precaution.
+        if input_multiprocess_fd is not None:
+            sys.stdin = os.fdopen(input_multiprocess_fd._handle)
+
+        pkg = serialized_pkg.restore()
+
+        if not kwargs['fake']:
+            setup_package(pkg, dirty=kwargs['dirty'])
+        return_value = function(pkg, kwargs)
+        child_pipe.send(return_value)
+
+    except StopPhase as e:
+        # Do not create a full ChildError from this, it's not an error
+        # it's a control statement.
+        child_pipe.send(e)
+    except BaseException:
+        # catch ANYTHING that goes wrong in the child process
+        exc_type, exc, tb = sys.exc_info()
+
+        # Need to unwind the traceback in the child because traceback
+        # objects can't be sent to the parent.
+        tb_string = traceback.format_exc()
+
+        # build up some context from the offending package so we can
+        # show that, too.
+        package_context = get_package_context(tb)
+
+        build_log = None
+        try:
+            if hasattr(pkg, 'log_path'):
+                build_log = pkg.log_path
+        except NameError:
+            # 'pkg' is not defined yet
+            pass
+
+        # make a pickleable exception to send to parent.
+        msg = "%s: %s" % (exc_type.__name__, str(exc))
+
+        ce = ChildError(msg,
+                        exc_type.__module__,
+                        exc_type.__name__,
+                        tb_string, build_log, package_context)
+        child_pipe.send(ce)
+
+    finally:
+        child_pipe.close()
+        if input_multiprocess_fd is not None:
+            input_multiprocess_fd.close()
+
+
+class TransmitPackage(object):
```

Comment:
I can come up with a better name, but I think it's worth noting that outside of a test context, the only thing that needs to be transmitted is the package: all the other serialization (i.e. that goes on inside `TransmitTestState`) is only required for tests (because they modify configs typically read from disk in memory and expect the subprocess to reflect that).

Reply:
But that might not always be true, even if it is right now. I'd rather have a general method, that we can then "dogfood" by using it for our tests, than have a special case for passing tests.